isobus/src/can_network_manager.cpp

```diff
@@ -364,13 +395,33 @@ namespace isobus
 		return retVal;
 	}
 
-	CANNetworkManager::CANNetworkManager() :
-	  updateTimestamp_ms(0),
-	  initialized(false)
+	CANNetworkManager::CANNetworkManager()
 	{
+		currentBusloadBitAccumulator.fill(0);
 		controlFunctionTable.fill({ nullptr });
 	}
 
+	std::uint32_t CANNetworkManager::get_number_bits_in_message(std::uint8_t numberDataBytes, CANIdentifier::Type messageType)
+	{
+		constexpr std::uint32_t MAX_CONSECUTIVE_SAME_BITS = 5; // After 5 consecutive bits, 6th will be opposite
+		const std::uint32_t dataLengthBits = CAN_DATA_LENGTH * numberDataBytes;
+		std::uint32_t retVal = 0;
+
+		if (CANIdentifier::Type::Extended == messageType)
+		{
+			constexpr std::uint32_t EXTENDED_ID_BEST_NON_DATA_LENGTH = 67; // SOF, ID, Control, CRC, ACK, EOF, and interframe space
+			constexpr std::uint32_t EXTENDED_ID_WORST_NON_DATA_LENGTH = 78;
+			retVal = ((dataLengthBits + EXTENDED_ID_BEST_NON_DATA_LENGTH) + (dataLengthBits + (dataLengthBits / MAX_CONSECUTIVE_SAME_BITS) + EXTENDED_ID_WORST_NON_DATA_LENGTH));
+		}
+		else
+		{
+			constexpr std::uint32_t STANDARD_ID_BEST_NON_DATA_LENGTH = 47; // SOF, ID, Control, CRC, ACK, EOF, and interframe space
+			constexpr std::uint32_t STANDARD_ID_WORST_NON_DATA_LENGTH = 54;
+			retVal = ((dataLengthBits + STANDARD_ID_BEST_NON_DATA_LENGTH) + (dataLengthBits + (dataLengthBits / MAX_CONSECUTIVE_SAME_BITS) + STANDARD_ID_WORST_NON_DATA_LENGTH));
+		}
+		return retVal;
+	}
+
```

Comment:
The tricky part about that is, I use the `HardwareInterfaceCANFrame` for calculating bits on transmits, but I use the existing receive in the network manager for the Rx bits, which is a  `const CANMessage &` so this seemed like a middle ground to handle both types

Reply:
Hmm, I see. Though wouldn't using the CANMessage lead to a inaccurate busload calculations sometimes? For example when receiving a large message with ETP, then the CANMessage would have received a large amount of bits over a long period of time, right? We might be able to shift the RX update to `CANNetworkManager::process_receive_can_message_frame(const HardwareInterfaceCANFrame &rxFrame)` instead? This would solve both problems I think

Reply:
I think when a transport level protocol completes, it just calls the callbacks directly, bypassing the problematic situation you described. Each message that comprises the protocol session should get individually accounted for though.

Regardless, yeah it probably makes sense to make it consistent and do it with the frames themselves, so I will make a change as you described.

Reply:
Ready for your re-review whenever you have time.